lib/semantics/check-do.cc

```diff
@@ -219,6 +214,46 @@ class DoConcurrentBodyEnforce {
   }
 
 private:
+  void CheckAllocateObjectIsCoarray(
+      const parser::AllocateObject &allocateObject, const char *stmtType) {
+    const Symbol *symbol{nullptr};
+    parser::CharBlock position;
+    std::visit(
+        common::visitors{
+            [&](const parser::Name &x) {
+              symbol = x.symbol;
+              position = x.source;
+            },
+            [&](const parser::StructureComponent &x) {
+              symbol = x.component.symbol;
+              position = x.component.source;
+            },
+        },
+        allocateObject.u);
```

Comment:
`parser::GetLastName` that returns a `parse::Name&` should do something similar here and could hide the details of `parser::AllocateObject` here as well as avoid the lambda definition.

Reply:
Thanks, Jean!  ```GetLastName()``` works like a champ.

Reply:
Also, I changed the test to use nested types to exercise this.